File: src/monitoring/logsEventProcessor.ts
import { PublicKey } from '@solana/web3.js';
import { Buffer } from 'buffer';
import * as base58 from 'bs58'; // bs58 is a dependency of @solana/web3.js
import {
  pumpFunProgramId, // Assuming this is PumpAddresses.PROGRAM
  systemProgramId, // Assuming this is SystemAddresses.PROGRAM (though not directly used in Python's _findAssociatedBondingCurve)
  tokenProgramId, // SystemAddresses.TOKEN_PROGRAM
  associatedTokenProgramId // SystemAddresses.ASSOCIATED_TOKEN_PROGRAM
} from '../core/pubkeys'; // Adjust path as necessary

// Re-define TokenInfo or import if defined elsewhere (e.g. from blockEventProcessor)
// For now, defining it based on Python code's usage.
export interface TokenInfo {
  name: string;
  symbol: string;
  uri: string;
  mint: PublicKey;
  bondingCurve: PublicKey;
  associatedBondingCurve: PublicKey;
  user: PublicKey; // The account that initiated the create, likely the fee payer
  creator: PublicKey;
  creatorVault: PublicKey;
  signature?: string; // Optional: include signature for context
}

// Placeholder for Trader interface
interface Trader {}

const logger = {
  info: (...args: any[]) => console.log('[LogsEventProcessor]', ...args),
  warn: (...args: any[]) => console.warn('[LogsEventProcessor]', ...args),
  error: (...args: any[]) => console.error('[LogsEventProcessor]', ...args),
};

export class LogsEventProcessor {
  private pumpProgramId: PublicKey;
  // Discriminator for create instruction data found in "Program data:" logs
  // Value from Python: 8530921459188068891
  private static readonly CREATE_INSTRUCTION_LOG_DISCRIMINATOR_STR = "1b773b243937779c"; // 8530921459188068891 as little-endian hex
  private static readonly CREATE_INSTRUCTION_LOG_DISCRIMINATOR = Buffer.from(LogsEventProcessor.CREATE_INSTRUCTION_LOG_DISCRIMINATOR_STR, "hex");

  constructor(pumpProgramIdToUse: PublicKey = pumpFunProgramId) {
    this.pumpProgramId = pumpProgramIdToUse;
  }

  /**
   * Process program logs and extract token info if a 'Create' event is found.
   * This specifically looks for "Program data:" logs associated with a "Create" instruction.
   */
  public processLogs(logs: string[], signature: string): TokenInfo | null {
    // Check if this is likely a token creation log based on keywords
    if (!logs.some(log => log.includes("Program log: Instruction: Create"))) {
      return null;
    }
    // Further filter out other instructions that might also contain "Create" in their name but aren't new tokens
    if (logs.some(log => log.includes("Program log: Instruction: CreateTokenAccount"))) {
        // logger.info("Skipping log for CreateTokenAccount instruction.");
        return null;
    }
    // Add more negative keywords if needed, e.g., "CreateMetadataAccount"

    for (const log of logs) {
      if (log.startsWith("Program data: ")) {
        try {
          const encodedData = log.substring("Program data: ".length);
          const decodedData = Buffer.from(encodedData, 'base64');
          
          const parsedData = this.parseCreateInstructionDataFromLog(decodedData);

          if (parsedData && parsedData.name) { // Check for 'name' as a key indicator of successful parsing
            const mint = new PublicKey(parsedData.mint);
            const bondingCurve = new PublicKey(parsedData.bondingCurve);
            // Note: The Python code's _find_associated_bonding_curve uses bonding_curve as the owner for ATA.
            // This seems specific. Standard ATA is derived from (wallet_address, token_program_id, mint_address).
            // Here, it's (bonding_curve_address, token_program_id, mint_address).
            const associatedCurve = this.findAssociatedBondingCurve(mint, bondingCurve);
            const creator = new PublicKey(parsedData.creator);
            const creatorVault = this.findCreatorVault(creator);

            return {
              name: parsedData.name,
              symbol: parsedData.symbol,
              uri: parsedData.uri,
              mint: mint,
              bondingCurve: bondingCurve,
              associatedBondingCurve: associatedCurve,
              user: new PublicKey(parsedData.user),
              creator: creator,
              creatorVault: creatorVault,
              signature: signature,
            };
          }
        } catch (e: any) {
          logger.error(`Failed to process log data for signature ${signature}: ${e.message}`);
        }
      }
    }
    return null;
  }

  /**
   * Parses the "Create" instruction data found within "Program data:" logs.
   * This is based on the specific structure assumed by the Python code.
   * TODO: This manual parsing is brittle. If a borsh schema for this event data exists, use it.
   */
  private parseCreateInstructionDataFromLog(data: Buffer): { [key: string]: string } | null {
    if (data.length < 8) {
      // logger.warn("Data too short to contain discriminator.");
      return null;
    }

    const discriminator = data.subarray(0, 8);
    if (!discriminator.equals(LogsEventProcessor.CREATE_INSTRUCTION_LOG_DISCRIMINATOR)) {
      // logger.info(`Skipping non-Create instruction data with discriminator: ${discriminator.toString('hex')}`);
      return null;
    }

    let offset = 8;
    const parsedData: { [key: string]: string } = {};

    // Fields based on Python's LogsEventProcessor._parse_create_instruction
    const fields: Array<{ name: string; type: 'string' | 'publicKey' }> = [
      { name: "name", type: "string" },
      { name: "symbol", type: "string" },
      { name: "uri", type: "string" },
      { name: "mint", type: "publicKey" },
      { name: "bondingCurve", type: "publicKey" },
      { name: "user", type: "publicKey" }, // This is likely the fee payer / tx sender
      { name: "creator", type: "publicKey" }, // This seems to be the intended creator authority
    ];

    try {
      for (const field of fields) {
        if (offset >= data.length) {
          logger.error(`Buffer underrun while parsing ${field.name}. Offset: ${offset}, Data Length: ${data.length}`);
          return null;
        }
        if (field.type === "string") {
          if (offset + 4 > data.length) {
             logger.error(`Not enough data for string length: field ${field.name}`); return null;
          }
          const length = data.readUInt32LE(offset);
          offset += 4;
          if (offset + length > data.length) {
            logger.error(`Not enough data for string content: field ${field.name}, length ${length}`); return null;
          }
          parsedData[field.name] = data.subarray(offset, offset + length).toString('utf-8');
          offset += length;
        } else if (field.type === "publicKey") {
           if (offset + 32 > data.length) {
            logger.error(`Not enough data for public key: field ${field.name}`); return null;
           }
          parsedData[field.name] = new PublicKey(data.subarray(offset, offset + 32)).toBase58();
          offset += 32;
        }
      }
      return parsedData;
    } catch (e: any) {
      logger.error(`Failed to parse create instruction data from log: ${e.message}`);
      return null;
    }
  }

  /**
   * Derives the Associated Token Account (ATA) for a given mint and owner.
   * In the context of the Python code, the 'bonding_curve' address was used as the owner.
   */
  private findAssociatedBondingCurve(mint: PublicKey, bondingCurveOwner: PublicKey): PublicKey {
    const [derivedAddress] = PublicKey.findProgramAddressSync(
      [bondingCurveOwner.toBuffer(), tokenProgramId.toBuffer(), mint.toBuffer()],
      associatedTokenProgramId
    );
    return derivedAddress;
  }

  private findCreatorVault(creator: PublicKey): PublicKey {
    const [derivedAddress] = PublicKey.findProgramAddressSync(
      [Buffer.from("creator-vault"), creator.toBuffer()],
      this.pumpProgramId // Use the stored pump program ID
    );
    return derivedAddress;
  }

  // --- Placeholder methods as per subtask description ---

  /**
   * TODO: Implement logic to parse buy transaction details from logs.
   * The structure of logs for buy transactions needs to be analyzed.
   * This might involve looking for specific program log messages or CPI calls.
   */
  public parseBuyTx(logs: string[], signature: string): any | null {
    logger.warn(`parseBuyTx for ${signature} not yet implemented. Logs:`, logs.join('\n'));
    // Example: Look for logs like "Program log: Instruction: Buy"
    // Then parse relevant data from associated "Program data:" or other logs.
    return null;
  }

  /**
   * TODO: Implement logic to parse sell transaction details from logs.
   * The structure of logs for sell transactions needs to be analyzed.
   */
  public parseSellTx(logs: string[], signature: string): any | null {
    logger.warn(`parseSellTx for ${signature} not yet implemented. Logs:`, logs.join('\n'));
    // Example: Look for logs like "Program log: Instruction: Sell"
    return null;
  }
}

File: src/monitoring/logs_event_processor.py
"""
Event processing for pump.fun tokens using logsSubscribe data.
"""

import base64
import struct
from typing import Final

import base58
from solders.pubkey import Pubkey

from core.pubkeys import PumpAddresses, SystemAddresses
from trading.base import TokenInfo
from utils.logger import get_logger

logger = get_logger(__name__)


class LogsEventProcessor:
    """Processes events from pump.fun program logs."""

    # Discriminator for create instruction to avoid non-create transactions
    CREATE_DISCRIMINATOR: Final[int] = 8530921459188068891

    def __init__(self, pump_program: Pubkey):
        """Initialize event processor.

        Args:
            pump_program: Pump.fun program address
        """
        self.pump_program = pump_program

    def process_program_logs(self, logs: list[str], signature: str) -> TokenInfo | None:
        """Process program logs and extract token info.

        Args:
            logs: List of log strings from the notification
            signature: Transaction signature

        Returns:
            TokenInfo if a token creation is found, None otherwise
        """
        # Check if this is a token creation
        if not any("Program log: Instruction: Create" in log for log in logs):
            return None

        # Skip swaps as the first condition may pass them
        if any("Program log: Instruction: CreateTokenAccount" in log for log in logs):
            return None

        # Find and process program data
        for log in logs:
            if "Program data:" in log:
                try:
                    encoded_data = log.split(": ")[1]
                    decoded_data = base64.b64decode(encoded_data)
                    parsed_data = self._parse_create_instruction(decoded_data)

                    if parsed_data and "name" in parsed_data:
                        mint = Pubkey.from_string(parsed_data["mint"])
                        bonding_curve = Pubkey.from_string(parsed_data["bondingCurve"])
                        associated_curve = self._find_associated_bonding_curve(
                            mint, bonding_curve
                        )
                        creator = Pubkey.from_string(parsed_data["creator"])
                        creator_vault = self._find_creator_vault(creator)

                        return TokenInfo(
                            name=parsed_data["name"],
                            symbol=parsed_data["symbol"],
                            uri=parsed_data["uri"],
                            mint=mint,
                            bonding_curve=bonding_curve,
                            associated_bonding_curve=associated_curve,
                            user=Pubkey.from_string(parsed_data["user"]),
                            creator=creator,
                            creator_vault=creator_vault,
                        )
                except Exception as e:
                    logger.error(f"Failed to process log data: {e}")

        return None

    def _parse_create_instruction(self, data: bytes) -> dict | None:
        """Parse the create instruction data.

        Args:
            data: Raw instruction data

        Returns:
            Dictionary of parsed data or None if parsing fails
        """
        if len(data) < 8:
            return None

        # Check for the correct instruction discriminator
        discriminator = struct.unpack("<Q", data[:8])[0]
        if discriminator != self.CREATE_DISCRIMINATOR:
            logger.info(
                f"Skipping non-Create instruction with discriminator: {discriminator}"
            )
            return None

        offset = 8
        parsed_data = {}

        # Parse fields based on CreateEvent structure
        fields = [
            ("name", "string"),
            ("symbol", "string"),
            ("uri", "string"),
            ("mint", "publicKey"),
            ("bondingCurve", "publicKey"),
            ("user", "publicKey"),
            ("creator", "publicKey"),
        ]

        try:
            for field_name, field_type in fields:
                if field_type == "string":
                    length = struct.unpack("<I", data[offset : offset + 4])[0]
                    offset += 4
                    value = data[offset : offset + length].decode("utf-8")
                    offset += length
                elif field_type == "publicKey":
                    value = base58.b58encode(data[offset : offset + 32]).decode("utf-8")
                    offset += 32

                parsed_data[field_name] = value

            return parsed_data
        except Exception as e:
            logger.error(f"Failed to parse create instruction: {e}")
            return None

    def _find_associated_bonding_curve(
        self, mint: Pubkey, bonding_curve: Pubkey
    ) -> Pubkey:
        """
        Find the associated bonding curve for a given mint and bonding curve.
        This uses the standard ATA derivation.

        Args:
            mint: Token mint address
            bonding_curve: Bonding curve address

        Returns:
            Associated bonding curve address
        """
        derived_address, _ = Pubkey.find_program_address(
            [
                bytes(bonding_curve),
                bytes(SystemAddresses.TOKEN_PROGRAM),
                bytes(mint),
            ],
            SystemAddresses.ASSOCIATED_TOKEN_PROGRAM,
        )
        return derived_address

    def _find_creator_vault(self, creator: Pubkey) -> Pubkey:
        """
        Find the creator vault for a creator.

        Args:
            creator: Creator address

        Returns:
            Creator vault address
        """
        derived_address, _ = Pubkey.find_program_address(
            [b"creator-vault", bytes(creator)],
            PumpAddresses.PROGRAM,
        )
        return derived_address
